Move schema loading/validating code out of config

jdavcs opened the pull request:
This replaces #9307. 

Motivation: `config/__init__.py` is doing too much. Last year's modifications to the config system, while simplifying configuration loading, have done the opposite to the configuration module's code, and especially to the `GalaxyAppConfiguration` class. This is an attempt to address this problem by separating 2 different responsibilities of the `GalaxyAppConfiguration` class: (1) automatically loading and processing all config properties from the schema; and (2) setting individual config properties and handling special cases. My first attempt (#9307) was to delegate the former to a ConfigurationLoader module and use it for any further schema processing, while using GalaxyAppConfiguration for setting individual config properties (as it did before).

However, concerns were raised during the code review: (a) the schema loader object was manipulating the internal structure of the config object; and (b) such tight coupling of 2 objects suggested the separation may have been not necessary (https://github.com/galaxyproject/galaxy/pull/9307#pullrequestreview-356385467). 

This, I think, is a better approach. Instead of delegating everything to a new "builder" object, I've split up the stuff that IMO does not belong in `GalaxyAppConfiguration` into 2 parts: (a) schema-related; and (b) configuration-related. All schema-related functionality has been moved to the schema object: loading collections of schema data + validating that data are now the schema object's responsibilities. All the configuration-related stuff that deals with all attributes + their processing, and does *not* include setting individual, galaxy-specific properties, has been moved up to the base config class.

I've updated the tests; I've also simplified the mocking in the unit tests in `test/unit/config`. 

I expect this will simplify the config code and make modifications easier. Also, as a side benefit, we can now eliminate much redundant code from TS's config (e.g., #9379, if needed).

Ping @nsoranzo, @bgruening.  

(EDIT: more details in config messages.)

jdavcs committed:
Move schema loading/validating code out of config

Add integration test; update several unit tests

nsoranzo commented on lib/galaxy/config/schema.py:
```diff
@@ -33,8 +37,68 @@ def __init__(self, schema_path, app_name):
         app_schema = self.raw_schema['mapping'][app_name]['mapping']
-        super(AppSchema, self).__init__(app_schema)
+        self._preprocess(app_schema)
+        super().__init__(app_schema)
```
This syntax is Python3-only, I don't think we are at that stage yet.

jdavcs replied:
I wasn't sure how Python3-only our code could be. I'll keep it Python2-compatible.

jdavcs committed:
Make schema module python2-compatible; cleanup comments

jdavcs committed:
Refactor config unit tests

1. Modify test setup code to adjust to edits in previous commit. Do NOT
   change the tests.

2. Simplify tests; get rid of mocking where possible (use real objects).

jdavcs committed:
Move OPTION_DEFAULTS to Schema

Rationale: group related items together. Constants OPTION_DEFAULTS and
UNKNOWN_OPTION are closely related and are used in the same context.

OPTION_DEFAULTS is only used in `schema.py` and `config_manage.py`; the
latter imports items from `schema.py`. Thus, it is reasonable to place
both constants in the same module.

(cherry-picked from closed #9307)

jdavcs committed:
Log message instead of raising an exception

This commit drops the exception that guarded against accidental
overwriting of existing attributes, and instead ignores the setattr
call, logging a debug message (which was initialy proposed in #8728).

Rationale: with the previous version, base configuration properties
(config_dir, data_dir, etc.) had to be set *after* this method was
called (otherwise an exeption was raised), but *before* any path
resolving code (because these properties are used for path
resolution). As a result, config loading required a rigid sequence of steps:
1: load defaults and kwargs, create all config attributes.
2: set base configuration properties [CANNOT MOVE THIS]
3: resolve paths, do the rest: load individual properties, etc.

However, setting base configuration properties is a special case that is
*independent of schema loading*, is handled separately (in
set_config_base), and takes precendence over schema loading procedures.

This modification makes config loading more flexible, and logical:
1. set base configuration properties
2. do the rest (now any refactoring can go here)

(cherry-picked from closed #9307)

jdavcs committed:
Remove appconfig instance var; simplify code

GalaxyAppConfiguration has a reference to an AppSchema object (`schema`
instance variable), which provides access to relevant schema data. So we
can simplify:
`self.appschema[key].get('path_resolves_to')` >> `self.schema.paths_to_resolve.get(key)`

Accessing the application schema mapping directly is straightforward:
`self.schema.app_schema`

jdavcs committed:
Fix mock + related code

Add schema var to galaxy mock as a MockSchema stub object

jdavcs committed:
Move a lot of code from config into base class

Provide stub methods for TS config implementation as a temporary
measure. No refactoring of TS config in this commit.

jdavcs committed:
Change handling of deprecated_dirs

This will prevent an error in TS if the base config class
`_update_raw_config_from_kwargs()` method is invoked.

jdavcs committed:
Reorder base config code, edit comments

jdavcs committed:
Correct _load_schema method signature; edit comments

Following the principle of least surprise.
The stub methods are not invoked because the parent's __init__() method
is not called. They can be safely removed. However, if a future commit adds
a call to parent's __init__, without these stub methods the parent's
methods will be called, which may result in a surprise (e.g., attributes
would be created for all schema properties).

jdavcs committed:
Fix _deprecated_dirs method signature

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -162,6 +172,78 @@ def _set_config_directories():
+        # Override in subclasses (optional)
+        # {KEY: config option, VALUE: deprecated directory name}
+        # If VALUE == first directory in a user path that resolves to KEY, it will be stripped from the path
+        return None
```
I'd keep `deprecated_dirs` as a dict, which here in `BaseAppConfiguration` would just be `{}`.

jdavcs replied:
Done.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -100,11 +100,21 @@
+        self._raw_config = self.schema.defaults.copy()  # Save schema defaults as initial config values (raw_config)
+        self._update_raw_config_from_kwargs(kwargs)  # Overwrite raw_config with values passed in kwargs
+        self._create_attributes_from_raw_config()  # Create attributes based on raw_config
+        self._resolve_paths(self.schema.paths_to_resolve)  # Overwrite attribute values with resolved paths
```
No point in passing `self.schema.paths_to_resolve` as a parameter since it's an internal method.

jdavcs replied:
Yes, that was an oversight. Fixing.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -162,6 +172,78 @@ def _set_config_directories():
+            if not hasattr(self, key):
+                setattr(self, key, value)
+            else:
+                log.debug("Attribute '%s' is set and cannot be overwritten with value '%s'" % (key, value))
```
Is it safe to just `log.debug()` in this case? It was raising a `ConfigurationError` exception before.

jdavcs replied:
In my opinion, it is not only safe but also necessary. Here're my thoughts: (commit message): https://github.com/galaxyproject/galaxy/pull/9371/commits/777379729198a893519ad58f167be761a4ea867d 
And here's our original discussion (you and I made it an exception in the first place): https://github.com/galaxyproject/galaxy/pull/8728#pullrequestreview-295263870
(I unresolved that discussion so it's not hidden)

natefoo replied:
Seems this should at least be `log.warning()` then? How would the attrs be created prior to calling this method?

jdavcs replied:
The following are legit cases when this would happen: `config_dir`, `data_dir`, `managed_data_dir` - i.e., the properties that are set in the base class' `_set_config_base()` method **and** defined in the schema. In the previous setup they had to be defined *after* this method was called but *before* any path resolution, as they were used in paths. So there was a rigid sequence of method calls: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L224 (see lines 224-230). Such tight coupling, of course, prevents refactoring (and is bad for other reasons).

I think the setting of base config files is a special case and has nothing to do with automatically reading schema and values from kwargs ( here's my thoughts: https://github.com/galaxyproject/galaxy/commit/777379729198a893519ad58f167be761a4ea867d ). And so considering these legit cases when this can happen, I think we can treat it as normal behavior: create attribute unless it exists. 

Besides, I think, the only time that can bite someone, is when you add a new instance variable to the config object and use a name for it that's in the schema (or vise-versa). But that's a remote possibility; besides, it's easily debuggable, especially considering the `create_attr...` method is called so early: https://github.com/galaxyproject/galaxy/blob/163ba59bb189cbb3bc42c7f7404ee075a6da8aeb/lib/galaxy/config/__init__.py#L118 - anything conflicting with this would have to be created *earlier* than that call. In summary, I think treating it as expected behavior is reasonable.

jdavcs replied:
Scratch all the above. See latest commit - just check the key against 3 base config attrs and raise exception for anything else.

natefoo replied:
Does it need an `else: log.something()` rather than silently passing now?

jdavcs replied:
I don't think so: it's expected behavior and happens on every load:
```
if attribute doesn't exist:
    create attribute - as expected 
else:
    if exists and is NOT a base config:
        raise error
    else: (is a base config)
        do not create attribute - as expected
```

natefoo replied:
Ok, thanks for the clarification

jdavcs committed:
Cleanup (address items from code review)

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -100,11 +100,24 @@
 class BaseAppConfiguration(object):
+    # Override in subclasses (optional): {KEY: config option, VALUE: deprecated directory name}
+    # If VALUE == first directory in a user-supplied path that resolves to KEY, it will be stripped from that path
+    deprecated_dirs = None
```
```suggestion
    deprecated_dirs = {}
```

natefoo replied:
Isn't this still relevant? If the subclass doesn't override, the `.get()`s on this below will raise an exception.

jdavcs replied:
It's only called here: https://github.com/galaxyproject/galaxy/blob/163ba59bb189cbb3bc42c7f7404ee075a6da8aeb/lib/galaxy/config/__init__.py#L210 - and the code checks if the variable has been set. I don't foresee any other use cases for it, so I think `deprecated_dirs = None` is a more straightforward way to say "you ain't gonna need it, unless you know you need it". Whereas an empty dict implies the possibility of loading stuff into it. 
That said, if you think such a safeguard is a reasonable precaution, I'll make it a dict as per Nicola's suggesion.

natefoo replied:
Here as well: https://github.com/galaxyproject/galaxy/pull/9371/files#diff-46f86ea0d82b84642e4218279ef61442R196 but I see that's covered by the code you linked.

jdavcs replied:
Yep, that's the one I was referring to.

natefoo commented on lib/galaxy/tools/toolbox/base.py:
```diff
@@ -1217,9 +1217,10 @@ def _init_dependency_manager(self):
-        self.dependency_manager = build_dependency_manager(app_config_dict=app_config_dict, conf_file=conf_file,
-                                                           default_tool_dependency_dir=default_tool_dependency_dir)
+        default_tool_dependency_dir = os.path.join(
+            self.app.config.data_dir, self.app.config.schema.app_schema['tool_dependency_dir'].get('default'))
```
What is the distinction between `config.schema.defaults['foo']` and `config.schema.app_schema['foo'].get('default')`?

jdavcs replied:
@natefoo None at all! It makes sense to use it, of course; but the mock schema has no property `defaults` - so toolbox unit tests will fail. It's a trade-off between adding some minor technical dept here, and modifying galaxy mock (`MockAppConfig`, which has a definition for `MockSchema`, which is a `Bunch` and doesn't have a `defaults` property). etc.

natefoo replied:
Ah ok. It'd be fairly difficult for folks to know whether it's safe to use `config.schema.defaults` in a given scenario then, though?

jdavcs replied:
Yes, indeed.. It's supposed to simplify, not to add a special case. I'll look into modifying the schema mock object to accommodate this. Leaving this open/unresolved for now.

jdavcs replied:
This, too, proved to be annoyingly simple. See last commit.

natefoo commented on lib/tool_shed/webapp/config.py:
```diff
@@ -34,6 +38,18 @@ class ConfigurationError(Exception):
+    def _load_schema(self):
+        return AppSchema(TOOLSHED_CONFIG_SCHEMA_PATH, TOOLSHED_APP_NAME)
+
+    def _update_raw_config_from_kwargs(self, kwargs):
+        pass  # Intentionally does nothing. To enable, remove + call base __init__().
+
+    def _create_attributes_from_raw_config(self):
+        pass  # Intentionally does nothing. To enable, remove + call base __init__().
+
+    def _resolve_paths(self, paths_to_resolve):
+        pass  # Intentionally does nothing. To enable, remove + call base __init__().
+
```
Why doesn't the TS use these?

jdavcs replied:
Oh, but it's about to :-) https://github.com/galaxyproject/galaxy/pull/9379

natefoo reviewed (approved):
Now I get what you were telling me about using DAGs, that part is pretty neat. I think overall the structure of this looks pretty good.

natefoo commented:
I unresolved a couple resolved comments that I added comments to

jdavcs committed:
Raise exception when trying to overwrite existing attr

Except base config attributes set in `_set_config_base()`

jdavcs committed:
Use new schema.defaults property

Instead of app_schema['foo'].get('default') - as per code review

jdavcs commented:
Thank you for the detailed review @natefoo!

bgruening commented:
Ah, sorry, missed that one. This is cool! Thanks @ic4f!

jdavcs commented:
@bgruening No worries, and thank you!
